crates/fuel-core/src/graphql_api/indexation/coins_to_spend.rs

```diff
@@ -0,0 +1,767 @@
+use fuel_core_storage::StorageAsMut;
+
+use fuel_core_types::{
+    entities::{
+        coins::coin::Coin,
+        Message,
+    },
+    fuel_tx::AssetId,
+    services::executor::Event,
+};
+
+use crate::graphql_api::{
+    ports::worker::OffChainDatabaseTransaction,
+    storage::coins::{
+        CoinsToSpendIndex,
+        CoinsToSpendIndexKey,
+        IndexedCoinType,
+    },
+};
+
+use super::error::IndexationError;
+
+// Indicates that a message is retryable.
+pub(crate) const RETRYABLE_BYTE: [u8; 1] = [0x00];
+
+// Indicates that a message is non-retryable (also, all coins use this byte).
+pub(crate) const NON_RETRYABLE_BYTE: [u8; 1] = [0x01];
+
+fn add_coin<T>(block_st_transaction: &mut T, coin: &Coin) -> Result<(), IndexationError>
+where
+    T: OffChainDatabaseTransaction,
+{
+    let key = CoinsToSpendIndexKey::from_coin(coin);
+    let storage = block_st_transaction.storage::<CoinsToSpendIndex>();
+    let maybe_old_value = storage.replace(&key, &IndexedCoinType::Coin)?;
+    if maybe_old_value.is_some() {
+        return Err(IndexationError::CoinToSpendAlreadyIndexed {
```

Comment:
If we return an error here the coin will still be inserted in the transaction and not revert. If we drop the transaction in an outer function it might not be an issue but maybe the dev of the outer function isn't aware of this. Same for `add_messages`

Reply:
Yes, there is an entire issue that tracks the missing error handling for indexation: https://github.com/FuelLabs/fuel-core/issues/2428 (referenced [here](https://github.com/FuelLabs/fuel-core/pull/2463/files#diff-1ac4adf5631afae925791c3897c52c2fa6282bd70901fddf552e7e1f0492c447R225)).

It's a little bit tricky, since these errors should never happen in sunny-day flow. If they happen, it means that the index got corrupted and we should probably disable them and fallback to legacy handling of the queries. Such corruption is fatal and cannot be recovered from, so we'd need to investigate how it happened. We might also consider advising the admin (via log) to re-sync the node from scratch to re-enable indexing.

But yeah, this is to be discussed/designed in the follow-up issue.